xdp-document.c

```diff
@@ -428,6 +462,48 @@ xdp_document_handle_finish_update (XdpDocument *doc,
   document_update_free (update);
 }
 
+static void
+xdp_document_handle_grant_permissions (XdpDocument *doc,
+                                       GDBusMethodInvocation *invocation,
+                                       const char *app_id,
+                                       GVariant *parameters)
+{
+  const char *target_app_id;
+  char **permissions;
+  XdpPermissionFlags perms;
+  gint i;
+
+  g_variant_get (parameters, "(&s^a&s)", &target_app_id, &permissions);
+
+  if (!xdp_document_has_permissions (doc, app_id, XDP_PERMISSION_FLAGS_GRANT_PERMISSIONS))
+    {
+      g_dbus_method_invocation_return_error (invocation, XDP_ERROR, XDP_ERROR_FAILED,
+                                             "No permissions to grant permissions");
+      return;
+    }
+
+  perms = 0;
+  for (i = 0; permissions[i]; i++)
+    {
+      if (strcmp (permissions[i], "read") == 0)
+        perms |= XDP_PERMISSION_FLAGS_READ;
+      else if (strcmp (permissions[i], "write") == 0)
+        perms |= XDP_PERMISSION_FLAGS_WRITE;
+      else if (strcmp (permissions[i], "grant-permissions") == 0)
+        perms |= XDP_PERMISSION_FLAGS_GRANT_PERMISSIONS;
+      else
+        {
+          g_dbus_method_invocation_return_error (invocation, XDP_ERROR, XDP_ERROR_FAILED,
+                                                 "No such permission: %s", permissions[i]);
+          return;
+        }
+    }
+
+  xdp_document_grant_permissions (doc, target_app_id, perms);
```

Comment:
Also, we should not return from the grant method until the new permission is stable on disk.